document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/transformer/PostgresUnnestQueryTransformer.java

```diff
@@ -0,0 +1,267 @@
+ *
+ * <p>As an example, { "item":"Soap", "sales":[ { "city":"delhi" } ] }. If we try to access
+ * `sales.city` it will return null as there is no key at `sales`.
+ *
+ * <p>However, as mongo supports -EQ- operator on array field that behaves as contains internally,
+ * `sales.city-EQ-delhi` is a valid filter expression. Ideally, the client should have expressed
+ * contains as `sales-CONTAINS-{"city":"delhi"}`.
+ *
+ * <p>To support compatibility, this pre-processor moves any applied filter on the array field to
+ * unnest expression filter.
+ */
+public class PostgresUnnestQueryTransformer implements QueryTransformer {
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public Query transform(Query query) {
+    // Don't do any thing if there is no unnest expression
+    if (query.getFromTypeExpressions().size() <= 0) {
+      return query;
+    }
+
+    // Prepare an AND tree of filters. If, any OR filter, the full sub-tree needs to move.
+    // [x, y, z, OR, OR]
+    AndFiltersCollector andFiltersCollector = new AndFiltersCollector();
+    List<FilterTypeExpression> andFilters =
+        query
+            .getFilter()
+            .map(
+                filterTypeExpression ->
+                    filterTypeExpression.<List<FilterTypeExpression>>accept(andFiltersCollector))
+            .orElse(List.of());
+
+    // check if any filter matches with unnest expression
+    // if matches, move the matched one to unnest filter and remove it from main filter list
+    List<FilterTypeExpression> finalAndFilters = new ArrayList<>();
+    List<FromTypeExpression> finalFromTypeExpressions =
+        matchUnnestExpressionsWithFilters(
+            query.getFromTypeExpressions(), andFilters, finalAndFilters);
+
+    // return same query if no change
+    if (andFilters.size() == finalAndFilters.size()) return query;
+
+    // return modified query if change
+    TransformedQueryBuilder transformedQueryBuilder = new TransformedQueryBuilder(query);
+    Optional<FilterTypeExpression> finalFilter = prepareFinalAndFiltersChain(finalAndFilters);
+    transformedQueryBuilder.setFromClauses(finalFromTypeExpressions);
+    if (finalFilter.isPresent()) {
+      transformedQueryBuilder.setFilter(finalFilter.get());
+    } else {
+      transformedQueryBuilder.clearFilter();
+    }
+    return transformedQueryBuilder.build();
+  }
+
+  private Optional<FilterTypeExpression> prepareFinalAndFiltersChain(
+      List<FilterTypeExpression> finalFilters) {
+    return Optional.ofNullable(
+        finalFilters.stream().reduce(null, PostgresUnnestQueryTransformer::buildAndFilter));
+  }
+
+  private static FilterTypeExpression buildAndFilter(
+      FilterTypeExpression left, FilterTypeExpression right) {
+    if (left == null) return right;
+    if (right == null) return left;
+    return LogicalExpression.builder()
+        .operator(LogicalOperator.AND)
+        .operand(left)
+        .operand(right)
+        .build();
```

Comment:
Nit: Can be simplified to `and(left, right)`

Reply:
Done.